headingoffset-polyfill.js

```diff
@@ -1,157 +1,268 @@
   const headings = Array.from(
     container.querySelectorAll("h1, h2, h3, h4, h5, h6")
   );
-  for (const heading of headings) {
-    applyHeadingOffset(heading);
+  // Add headings in shadow roots
+  for (const element of container.querySelectorAll("*")) {
+    if (element.shadowRoot) {
+      /** @type {HTMLHeadingElement[]} */
+      const shadowHeadings = Array.from(
+        element.shadowRoot.querySelectorAll("h1, h2, h3, h4, h5, h6")
+      );
+      headings.push(...shadowHeadings);
+    }
   }
-}
-
-/** Track headings whose 'aria-level' is set by the polyfill. */
-function isManagedHeading(heading) {
-  return managedHeadings.has(heading);
-}
-function startManagingHeading(heading) {
-  managedHeadings.add(heading);
-}
-function stopManagingHeading(heading) {
-  managedHeadings.delete(heading);
+  return headings;
 }
 
 /**
- * Applies an offset to a heading, if needed.
- * @param {Element} heading A heading element.
+ * Updates a heading’s 'aria-level' attribute, based on its offset
+ * @param {HTMLHeadingElement} heading
  */
-function applyHeadingOffset(heading) {
-  // If 'aria-level' is already set and the polyfill didn’t set it, don’t change it.
-  if (heading.hasAttribute("aria-level") && !isManagedHeading(heading)) {
+function updateAriaLevel(heading) {
+  // Early-return if 'aria-level' was not set by the polyfill
+  if (heading.hasAttribute("aria-level") && !managedHeadings.has(heading)) {
     return;
   }
 
-  const level = Number(heading.tagName[1]);
-  const offset = getHeadingOffset(heading, 9 - level);
-  const ariaLevel = level + offset;
+  const tagLevel = Number(heading.tagName[1]);
+  const offset = computeHeadingOffset(heading, 9 - tagLevel);
+  const ariaLevel = tagLevel + offset;
 
-  // If the level wouldn’t change, don’t set 'aria-level', and remove it if it’s redundant.
-  if (ariaLevel === level) {
+  // Don’t set 'aria-level' if it’s redundant
+  if (ariaLevel === tagLevel) {
     heading.removeAttribute("aria-level");
     return;
   }
 
   heading.setAttribute("aria-level", String(ariaLevel));
-  ariaLevelObserver.takeRecords();
-  startManagingHeading(heading);
+  ariaLevelObserver.takeRecords(); // Prevent observer loop
+  managedHeadings.add(heading);
 }
 
 /**
- * Determines the number of levels to offset a heading’s level, by summing the 'headingoffset' attributes of all its parents.
- * If the offset would be greater than the maximum offset, returns 0.
+ * Determines whether an element is explicitly (via the 'headingreset' attribute)
+ * or implicitly (as a modal dialog) a 'headingoffset' accumulation boundary.
+ * @param {Element} element
+ * @returns {boolean}
+ */
+function isOffsetBoundary(element) {
+  // Modal dialogs implicitly reset heading levels
+  if (element instanceof HTMLDialogElement) {
+    return (
+      element.open &&
+      element.hasAttribute("open") &&
+      (element.matches(":modal") ||
+        element.ownerDocument.querySelector(":modal") === element)
```

Comment:
This feels redundant. I think it can be simply:

```suggestion
      element.matches(":modal")
```

If not, it might be worth trying to explain why.

Reply:
Thanks! I was adapting https://github.com/github/ariaNotify-polyfill/blob/4f1ba86d86df6e5fc5609bacdda6f6987ccf7269/ariaNotify-polyfill.js#L58-L63, but you’re right that this can be simplified, since here we aren’t checking ‘is this element _inside_ a modal dialog’ but rather ‘is this element a modal dialog’.